File: frontend/src/types/room-allocation.ts
// Room Allocation System Types
// Comprehensive type definitions for room allocation management

export type AssignmentType = 'auto' | 'manual' | 'guest_request' | 'upgrade' | 'downgrade';

export type AssignmentStatus = 'pre_assigned' | 'assigned' | 'locked' | 'cancelled';

export type BlockType = 'maintenance' | 'renovation' | 'vip_hold' | 'long_stay' | 'staff' | 'inspection' | 'deep_clean';

export type AlertType = 
  | 'unassigned_24h' 
  | 'unassigned_1h' 
  | 'unassigned_critical'
  | 'conflict_detected' 
  | 'upgrade_available' 
  | 'room_blocked'
  | 'assignment_failed' 
  | 'rate_change_required';

export type AlertSeverity = 'info' | 'warning' | 'critical' | 'emergency';

export type RoomStatus = 
  | 'available' 
  | 'occupied' 
  | 'blocked' 
  | 'maintenance' 
  | 'arriving' 
  | 'departing' 
  | 'pre_assigned';

export type AssignmentStrategy = 
  | 'optimize_occupancy' 
  | 'group_by_type' 
  | 'vip_first' 
  | 'distribute_wear';

// Core Allocation Interfaces
export interface RoomAllocation {
  id: string;
  booking_id: string;
  room_id: string;
  assignment_type: AssignmentType;
  assignment_status: AssignmentStatus;
  check_in_date: string;
  check_out_date: string;
  nights_count: number;
  original_room_type_id?: string;
  original_rate?: number;
  allocated_rate?: number;
  rate_difference?: number;
  assigned_at?: string;
  assigned_by?: string;
  assignment_reason?: string;
  previous_room_id?: string;
  changed_at?: string;
  changed_by?: string;
  change_reason?: string;
  guest_preferences?: Record<string, any>;
  internal_notes?: string;
  is_vip: boolean;
  is_guaranteed: boolean;
  requires_inspection: boolean;
  created_at: string;
  updated_at: string;
}

export interface CreateRoomAllocation {
  booking_id: string;
  room_id: string;
  assignment_type: AssignmentType;
  assignment_status: AssignmentStatus;
  check_in_date: string;
  check_out_date: string;
  assignment_reason?: string;
  guest_preferences?: Record<string, any>;
  internal_notes?: string;
  is_vip?: boolean;
  is_guaranteed?: boolean;
  requires_inspection?: boolean;
  original_room_type_id?: string;
  original_rate?: number;
  allocated_rate?: number;
  assigned_by?: string;
}

export interface UpdateRoomAllocation {
  room_id?: string;
  assignment_status?: AssignmentStatus;
  check_in_date?: string;
  check_out_date?: string;
  assignment_reason?: string;
  guest_preferences?: Record<string, any>;
  internal_notes?: string;
  is_guaranteed?: boolean;
  requires_inspection?: boolean;
  changed_by?: string;
  change_reason?: string;
}

// Assignment Requests
export interface AssignRoomRequest {
  booking_id: string;
  room_id: string;
  assignment_type?: AssignmentType;
  assignment_reason?: string;
  is_guaranteed?: boolean;
  override_conflicts?: boolean;
}

export interface AssignRoomResponse {
  allocation_id: string;
  room_assigned: {
    room_number: string;
    room_type: string;
    floor: number;
  };
  price_adjustment?: {
    original_rate: number;
    new_rate: number;
    difference: number;
    total_difference: number;
    nights: number;
  };
  status: AssignmentStatus;
  conflicts?: string[];
  warnings?: string[];
}

export interface AutoAssignRequest {
  date_range: {
    from: string;
    to: string;
  };
  assignment_strategy?: AssignmentStrategy;
  respect_preferences?: boolean;
  room_type_ids?: string[];
  include_vip_only?: boolean;
}

export interface AutoAssignResponse {
  assignments_created: number;
  assignments_failed: number;
  assignments: AssignRoomResponse[];
  failed_bookings: Array<{
    booking_id: string;
    booking_code: string;
    guest_name: string;
    reason: string;
    suggestions: string[];
  }>;
  summary: Record<string, any>;
}

export interface ChangeRoomRequest {
  new_room_id: string;
  change_reason: string;
  apply_charges?: boolean;
  notify_guest?: boolean;
  override_conflicts?: boolean;
}

export interface ChangeRoomResponse {
  allocation_id: string;
  previous_room: {
    room_number: string;
    room_type: string;
    floor: number;
  };
  new_room: {
    room_number: string;
    room_type: string;
    floor: number;
  };
  price_adjustment?: {
    original_rate: number;
    new_rate: number;
    difference: number;
    total_difference: number;
    nights: number;
  };
  charges_applied: boolean;
  status: string;
}

// Room Blocks
export interface RoomBlock {
  id: string;
  room_id: string;
  start_date: string;
  end_date: string;
  block_type: BlockType;
  block_reason?: string;
  is_active: boolean;
  can_override: boolean;
  override_level?: string;
  created_by?: string;
  created_at: string;
  released_at?: string;
  released_by?: string;
}

export interface CreateRoomBlock {
  room_id: string;
  start_date: string;
  end_date: string;
  block_type: BlockType;
  block_reason?: string;
  can_override?: boolean;
  override_level?: string;
  created_by?: string;
}

export interface ReleaseBlockRequest {
  release_reason: string;
  released_by?: string;
}

// Allocation Alerts
export interface AllocationAlert {
  id: string;
  booking_id: string;
  allocation_id?: string;
  alert_type: AlertType;
  alert_message: string;
  severity: AlertSeverity;
  alert_context?: Record<string, any>;
  is_resolved: boolean;
  resolved_at?: string;
  resolved_by?: string;
  resolution_notes?: string;
  auto_resolved: boolean;
  notified_users: string[];
  notification_channels: string[];
  notification_sent_at?: string;
  escalation_level: number;
  escalated_at?: string;
  escalated_to?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateAllocationAlert {
  booking_id: string;
  allocation_id?: string;
  alert_type: AlertType;
  alert_message: string;
  severity: AlertSeverity;
  alert_context?: Record<string, any>;
}

export interface UpdateAllocationAlert {
  is_resolved?: boolean;
  resolution_notes?: string;
  resolved_by?: string;
}

// Unassigned Bookings
export interface UnassignedBooking {
  id: string;
  booking_code: string;
  guest_name: string;
  guest_email?: string;
  guest_phone?: string;
  check_in_date: string;
  check_out_date: string;
  check_in_time?: string;
  room_type_id: string;
  room_type_name?: string;
  adults: number;
  children: number;
  nights: number;
  total_amount: number;
  booking_date: string;
  status: string;
  special_requests?: string;
  hours_until_checkin?: number;
  severity?: AlertSeverity;
  priority_score?: number;
  guest_preferences?: {
    floor_preference?: string;
    room_features?: string[];
    special_requirements?: string[];
  };
}

export interface UnassignedBookingsResponse {
  bookings: UnassignedBooking[];
  total: number;
  critical_count: number;
  warning_count: number;
  info_count: number;
  statistics: {
    total_unassigned: number;
    arriving_today: number;
    arriving_tomorrow: number;
    overdue: number;
    average_hours_until_checkin: number;
  };
}

// Monthly Grid Data
export interface RoomDailyStatus {
  date: string;
  status: RoomStatus;
  booking_id?: string;
  guest_name?: string;
  is_arrival: boolean;
  is_departure: boolean;
  is_vip: boolean;
  special_notes?: string;
  rate?: number;
  block_reason?: string;
}

export interface RoomGridData {
  room_id: string;
  room_number: string;
  room_type: string;
  floor: number;
  features: string[];
  daily_status: RoomDailyStatus[];
}

export interface MonthlyGridRequest {
  month: string; // Format: YYYY-MM
  room_type_ids?: string[];
  floors?: number[];
  include_blocked?: boolean;
  include_maintenance?: boolean;
}

export interface MonthlyGridResponse {
  month: string;
  days_in_month: number;
  rooms: RoomGridData[];
  summary: {
    total_rooms: number;
    days_in_month: number;
    average_occupancy_rate: number;
    peak_occupancy: number;
    total_bookings: number;
    active_blocks: number;
  };
  occupancy_stats: Array<{
    date: string;
    occupied: number;
    available: number;
    blocked: number;
    arriving: number;
    departing: number;
    occupancy_rate: number;
  }>;
}

// Unassigned Bookings
export interface UnassignedBooking {
  booking_id: string;
  booking_code: string;
  guest_name: string;
  phone?: string;
  email?: string;
  check_in_date: string;
  check_out_date: string;
  check_in_time?: string;
  check_out_time?: string;
  room_type: string;
  room_type_id: string;
  adults: number;
  children: number;
  infants: number;
  total_guests: number;
  extra_persons: number;
  extra_person_charge: number;
  extra_bed: boolean;
  extra_bed_charge: number;
  special_requests?: string;
  internal_notes?: any;
  is_vip: boolean;
  hours_until_checkin?: number;
  alert_level: AlertSeverity;
  total_amount: number;
  paid_amount: number;
  booking_status: string;
}

export interface UnassignedBookingsResponse {
  alerts: UnassignedBooking[];
  summary: {
    total_unassigned: number;
    critical: number;
    warning: number;
    info: number;
  };
  recommendations: string[];
}

// Bulk Operations
export interface BulkResolveAlertsRequest {
  alert_ids: string[];
  action: 'auto_assign' | 'manual_assign' | 'dismiss';
  notes?: string;
  assignment_strategy?: AssignmentStrategy;
}

export interface BulkResolveAlertsResponse {
  resolved_count: number;
  failed_count: number;
  results: Array<Record<string, any>>;
  errors: string[];
}

// Available Rooms
export interface AvailableRoomInfo {
  room_id: string;
  room_number: string;
  room_type_id: string;
  room_type: string;
  floor: number;
  features: string[];
  base_rate: number;
  current_status: string;
  availability_score?: number;
  is_accessible: boolean;
  last_cleaned?: string;
  maintenance_notes?: string;
}

export interface AvailableRoomsRequest {
  check_in_date: string;
  check_out_date: string;
  room_type_id?: string;
  guest_count?: number;
  accessibility_required?: boolean;
  features_required?: string[];
  exclude_rooms?: string[];
}

export interface AvailableRoomsResponse {
  available_rooms: AvailableRoomInfo[];
  total_available: number;
  filters_applied: Record<string, any>;
  suggestions: string[];
}

// Guest Preferences
export interface GuestRoomPreferences {
  id: string;
  customer_id: string;
  preferred_room_types: string[];
  preferred_floors: number[];
  preferred_features: string[];
  avoid_floors: number[];
  avoid_features: string[];
  avoid_rooms: string[];
  accessibility_needs: Record<string, any>;
  special_requests?: string;
  priority_level: number; // 1-10 scale
  last_stayed_room?: string;
  total_stays: number;
  satisfaction_score?: number; // 1-5 stars
  created_at: string;
  updated_at: string;
}

export interface CreateGuestPreferences {
  customer_id: string;
  preferred_room_types?: string[];
  preferred_floors?: number[];
  preferred_features?: string[];
  avoid_floors?: number[];
  avoid_features?: string[];
  avoid_rooms?: string[];
  accessibility_needs?: Record<string, any>;
  special_requests?: string;
  priority_level?: number;
}

export interface UpdateGuestPreferences {
  preferred_room_types?: string[];
  preferred_floors?: number[];
  preferred_features?: string[];
  avoid_floors?: number[];
  avoid_features?: string[];
  avoid_rooms?: string[];
  accessibility_needs?: Record<string, any>;
  special_requests?: string;
  priority_level?: number;
}

// Allocation Rules
export interface AllocationRule {
  id: string;
  rule_name: string;
  rule_type: string;
  conditions: Record<string, any>;
  actions: Record<string, any>;
  priority: number;
  is_active: boolean;
  created_by?: string;
  created_at: string;
  updated_at: string;
}

export interface CreateAllocationRule {
  rule_name: string;
  rule_type: string;
  conditions: Record<string, any>;
  actions: Record<string, any>;
  priority?: number;
  is_active?: boolean;
  created_by?: string;
}

export interface UpdateAllocationRule {
  rule_name?: string;
  conditions?: Record<string, any>;
  actions?: Record<string, any>;
  priority?: number;
  is_active?: boolean;
}

// History and Audit
export interface AllocationHistory {
  id: string;
  allocation_id: string;
  booking_id: string;
  action: string;
  previous_room_id?: string;
  new_room_id?: string;
  previous_room_number?: string;
  new_room_number?: string;
  previous_dates?: string;
  new_dates?: string;
  previous_status?: string;
  new_status?: string;
  price_adjustment?: number;
  changed_by?: string;
  change_reason?: string;
  changed_at: string;
}

// Analytics and Reports
export interface OptimizationMetrics {
  average_assignment_time: string;
  auto_assignment_rate: number;
  manual_assignment_rate: number;
  room_change_rate: number;
  unassigned_alerts: number;
  late_assignments: number;
}

export interface RoomUtilization {
  room: string;
  nights: number;
  utilization: number;
  revenue: number;
  avg_rate: number;
}

export interface OptimizationRecommendation {
  type: string;
  message: string;
  priority: string;
  action_required: boolean;
}

export interface OptimizationReport {
  period: {
    from: string;
    to: string;
  };
  metrics: OptimizationMetrics;
  room_utilization: {
    most_used_rooms: RoomUtilization[];
    least_used_rooms: RoomUtilization[];
  };
  recommendations: OptimizationRecommendation[];
  trends: Record<string, any>;
}

export interface AllocationStatistics {
  total_allocations: number;
  by_status: Record<AssignmentStatus, number>;
  by_type: Record<AssignmentType, number>;
  average_lead_time_hours: number;
  room_utilization_rate: number;
  conflict_resolution_time: number;
  guest_satisfaction_impact?: number;
}

export interface DailyAllocationSummary {
  date: string;
  total_arrivals: number;
  total_departures: number;
  rooms_assigned: number;
  rooms_available: number;
  occupancy_rate: number;
  unassigned_critical: number;
  unassigned_warning: number;
  conflicts_resolved: number;
}

// Dashboard Data
export interface AllocationDashboard {
  summary?: {
    total_rooms: number;
    occupied_rooms: number;
    available_rooms: number;
    occupancy_rate: number;
    unassigned_bookings: number;
  };
  unassigned_summary?: {
    total_unassigned: number;
    critical: number;
    warning: number;
    info: number;
  };
  today_arrivals: number | {
    count: number;
    list: Array<{
      booking_id: string;
      booking_code: string;
      guest_name: string;
      room_number: string;
      room_type: string;
      check_in_time: string;
    }>;
  };
  today_departures: number | {
    count: number;
    list: Array<{
      booking_id: string;
      booking_code: string;
      guest_name: string;
      room_number: string;
      room_type: string;
      check_out_time: string;
    }>;
  };
  alerts?: {
    unassigned_today: number;
    message: string;
  };
  total_rooms?: number;
  occupied_rooms?: number;
  available_rooms?: number;
  occupancy_rate?: number;
  recent_arrivals?: Array<{
    guest_name: string;
    room_number: string;
    time: string;
  }>;
  upcoming_departures?: Array<{
    guest_name: string;
    room_number: string;
    time: string;
  }>;
}

// Validation and Utility Types
export interface AssignmentValidation {
  valid: boolean;
  conflicts: string[];
  price_adjustment?: {
    original_rate: number;
    new_rate: number;
    difference: number;
    total_difference: number;
    nights: number;
  };
  room_details: {
    room_number: string;
    room_type: string;
    floor: number;
  };
}

// Grid View UI Types
export interface GridViewSettings {
  statusColors: {
    available: string;
    occupied: string;
    arriving: string;
    departing: string;
    blocked: string;
    maintenance: string;
    pre_assigned: string;
  };
  cellDisplay: {
    primary: 'guest_name' | 'booking_code' | 'status';
    secondary: 'nights_remaining' | 'room_rate';
    icons: {
      vip: string;
      long_stay: string;
      group: string;
      notes: string;
    };
  };
  interactions: {
    hover: string;
    click: string;
    drag_drop: string;
    right_click: string;
  };
}

// Filter and Search Types
export interface AllocationFilters {
  date_range?: {
    start: string;
    end: string;
  };
  assignment_status?: AssignmentStatus[];
  assignment_type?: AssignmentType[];
  room_types?: string[];
  floors?: number[];
  is_vip?: boolean;
  is_guaranteed?: boolean;
  requires_inspection?: boolean;
  search_term?: string;
}

export interface AlertFilters {
  severity?: AlertSeverity[];
  alert_type?: AlertType[];
  is_resolved?: boolean;
  date_range?: {
    start: string;
    end: string;
  };
  escalation_level?: number[];
}

// Component Props Types
export interface RoomGridCellProps {
  room: RoomGridData;
  date: string;
  status: RoomDailyStatus;
  isSelected?: boolean;
  onCellClick?: (room: RoomGridData, date: string) => void;
  onCellDoubleClick?: (room: RoomGridData, date: string) => void;
  onDragStart?: (room: RoomGridData, date: string) => void;
  onDrop?: (room: RoomGridData, date: string) => void;
}

export interface AllocationFormProps {
  booking?: any; // Booking object
  initialRoom?: AvailableRoomInfo;
  onSubmit: (allocation: CreateRoomAllocation) => void;
  onCancel: () => void;
  isLoading?: boolean;
}

export interface AlertListProps {
  alerts: AllocationAlert[];
  onResolve: (alertId: string, resolution: UpdateAllocationAlert) => void;
  onBulkResolve: (request: BulkResolveAlertsRequest) => void;
  filters?: AlertFilters;
  onFiltersChange?: (filters: AlertFilters) => void;
}

File: backend/app/schemas/room_allocation.py
from datetime import date, datetime, time
from decimal import Decimal
from typing import List, Optional, Dict, Any, Union
from uuid import UUID
from enum import Enum
from pydantic import BaseModel, Field, validator, root_validator


# Enums for validation
class AssignmentType(str, Enum):
    AUTO = "auto"
    MANUAL = "manual"
    GUEST_REQUEST = "guest_request"
    UPGRADE = "upgrade"
    DOWNGRADE = "downgrade"


class AssignmentStatus(str, Enum):
    PRE_ASSIGNED = "pre_assigned"
    ASSIGNED = "assigned"
    LOCKED = "locked"
    CANCELLED = "cancelled"


class BlockType(str, Enum):
    MAINTENANCE = "maintenance"
    RENOVATION = "renovation"
    VIP_HOLD = "vip_hold"
    LONG_STAY = "long_stay"
    STAFF = "staff"
    INSPECTION = "inspection"
    DEEP_CLEAN = "deep_clean"


class AlertType(str, Enum):
    UNASSIGNED_24H = "unassigned_24h"
    UNASSIGNED_1H = "unassigned_1h"
    UNASSIGNED_CRITICAL = "unassigned_critical"
    CONFLICT_DETECTED = "conflict_detected"
    UPGRADE_AVAILABLE = "upgrade_available"
    ROOM_BLOCKED = "room_blocked"
    ASSIGNMENT_FAILED = "assignment_failed"
    RATE_CHANGE_REQUIRED = "rate_change_required"


class AlertSeverity(str, Enum):
    INFO = "info"
    WARNING = "warning"
    CRITICAL = "critical"
    EMERGENCY = "emergency"


class RoomStatus(str, Enum):
    AVAILABLE = "available"
    OCCUPIED = "occupied"
    BLOCKED = "blocked"
    MAINTENANCE = "maintenance"
    ARRIVING = "arriving"
    DEPARTING = "departing"
    PRE_ASSIGNED = "pre_assigned"
    CLEANING = "cleaning"


class AssignmentStrategy(str, Enum):
    OPTIMIZE_OCCUPANCY = "optimize_occupancy"
    GROUP_BY_TYPE = "group_by_type"
    VIP_FIRST = "vip_first"
    DISTRIBUTE_WEAR = "distribute_wear"


# Base schemas
class RoomAllocationBase(BaseModel):
    booking_id: UUID
    room_id: UUID
    assignment_type: AssignmentType
    assignment_status: AssignmentStatus
    check_in_date: date
    check_out_date: date
    assignment_reason: Optional[str] = None
    guest_preferences: Optional[Dict[str, Any]] = {}
    internal_notes: Optional[str] = None
    is_vip: bool = False
    is_guaranteed: bool = False
    requires_inspection: bool = False

    @validator('check_out_date')
    def validate_checkout_after_checkin(cls, v, values):
        if 'check_in_date' in values and v <= values['check_in_date']:
            raise ValueError('Check-out date must be after check-in date')
        return v


class RoomAllocationCreate(RoomAllocationBase):
    original_room_type_id: Optional[UUID] = None
    original_rate: Optional[Decimal] = None
    allocated_rate: Optional[Decimal] = None
    assigned_by: Optional[UUID] = None


class RoomAllocationUpdate(BaseModel):
    room_id: Optional[UUID] = None
    assignment_status: Optional[AssignmentStatus] = None
    check_in_date: Optional[date] = None
    check_out_date: Optional[date] = None
    assignment_reason: Optional[str] = None
    guest_preferences: Optional[Dict[str, Any]] = None
    internal_notes: Optional[str] = None
    is_guaranteed: Optional[bool] = None
    requires_inspection: Optional[bool] = None
    changed_by: Optional[UUID] = None
    change_reason: Optional[str] = None


class RoomAllocationResponse(RoomAllocationBase):
    id: UUID
    nights_count: int
    original_room_type_id: Optional[UUID] = None
    original_rate: Optional[Decimal] = None
    allocated_rate: Optional[Decimal] = None
    rate_difference: Optional[Decimal] = None
    assigned_at: Optional[datetime] = None
    assigned_by: Optional[UUID] = None
    previous_room_id: Optional[UUID] = None
    changed_at: Optional[datetime] = None
    changed_by: Optional[UUID] = None
    change_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Room Assignment Request
class AssignRoomRequest(BaseModel):
    booking_id: UUID
    room_id: UUID
    assignment_type: AssignmentType = AssignmentType.MANUAL
    assignment_reason: Optional[str] = None
    is_guaranteed: bool = False
    override_conflicts: bool = False


class AssignRoomResponse(BaseModel):
    allocation_id: UUID
    room_assigned: Dict[str, Any]
    price_adjustment: Optional[Dict[str, Any]] = None
    status: AssignmentStatus
    conflicts: Optional[List[str]] = []
    warnings: Optional[List[str]] = []


# Auto Assignment Request
class AutoAssignRequest(BaseModel):
    date_range: Dict[str, date] = Field(..., description="from and to dates")
    assignment_strategy: AssignmentStrategy = AssignmentStrategy.OPTIMIZE_OCCUPANCY
    respect_preferences: bool = True
    room_type_ids: Optional[List[UUID]] = None
    include_vip_only: bool = False

    @validator('date_range')
    def validate_date_range(cls, v):
        if 'from' not in v or 'to' not in v:
            raise ValueError('Date range must contain "from" and "to" keys')
        if v['to'] < v['from']:
            raise ValueError('End date must be after start date')
        return v


class AutoAssignResponse(BaseModel):
    assignments_created: int
    assignments_failed: int
    assignments: List[AssignRoomResponse]
    failed_bookings: List[Dict[str, Any]]
    summary: Dict[str, Any]


# Room Change Request
class ChangeRoomRequest(BaseModel):
    new_room_id: UUID
    change_reason: str
    apply_charges: bool = True
    notify_guest: bool = False
    override_conflicts: bool = False


class ChangeRoomResponse(BaseModel):
    allocation_id: UUID
    previous_room: Dict[str, Any]
    new_room: Dict[str, Any]
    price_adjustment: Optional[Dict[str, Any]] = None
    charges_applied: bool
    status: str


# Room Blocks
class RoomBlockBase(BaseModel):
    room_id: UUID
    start_date: date
    end_date: date
    block_type: BlockType
    block_reason: Optional[str] = None
    can_override: bool = False
    override_level: Optional[str] = None

    @validator('end_date')
    def validate_end_after_start(cls, v, values):
        if 'start_date' in values and v < values['start_date']:
            raise ValueError('End date must be after start date')
        return v


class RoomBlockCreate(RoomBlockBase):
    created_by: Optional[UUID] = None


class RoomBlockUpdate(BaseModel):
    start_date: Optional[date] = None
    end_date: Optional[date] = None
    block_reason: Optional[str] = None
    can_override: Optional[bool] = None
    override_level: Optional[str] = None


class RoomBlockResponse(RoomBlockBase):
    id: UUID
    is_active: bool
    created_by: Optional[UUID] = None
    created_at: datetime
    released_at: Optional[datetime] = None
    released_by: Optional[UUID] = None

    class Config:
        from_attributes = True


class ReleaseBlockRequest(BaseModel):
    release_reason: str
    released_by: Optional[UUID] = None


# Allocation Alerts
class AllocationAlertBase(BaseModel):
    booking_id: UUID
    alert_type: AlertType
    alert_message: str
    severity: AlertSeverity
    alert_context: Optional[Dict[str, Any]] = {}


class AllocationAlertCreate(AllocationAlertBase):
    allocation_id: Optional[UUID] = None


class AllocationAlertUpdate(BaseModel):
    is_resolved: Optional[bool] = None
    resolution_notes: Optional[str] = None
    resolved_by: Optional[UUID] = None


class AllocationAlertResponse(AllocationAlertBase):
    id: UUID
    allocation_id: Optional[UUID] = None
    is_resolved: bool
    resolved_at: Optional[datetime] = None
    resolved_by: Optional[UUID] = None
    resolution_notes: Optional[str] = None
    auto_resolved: bool
    notified_users: List[UUID]
    notification_channels: List[str]
    notification_sent_at: Optional[datetime] = None
    escalation_level: int
    escalated_at: Optional[datetime] = None
    escalated_to: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Monthly Grid Data
class RoomDailyStatus(BaseModel):
    date: date
    status: RoomStatus
    booking_id: Optional[UUID] = None
    guest_name: Optional[str] = None
    is_arrival: bool = False
    is_departure: bool = False
    is_vip: bool = False
    special_notes: Optional[str] = None
    rate: Optional[Decimal] = None
    block_reason: Optional[str] = None
    # Shift-based booking fields
    shift_type: Optional[str] = None  # 'day_shift', 'night_shift', 'full_day', 'traditional'
    day_shift_booking: Optional[Dict[str, Any]] = None  # {'booking_id', 'guest_name', 'status'}
    night_shift_booking: Optional[Dict[str, Any]] = None  # {'booking_id', 'guest_name', 'status'}


class RoomGridData(BaseModel):
    room_id: UUID
    room_number: str
    room_type: str
    floor: int
    features: List[str] = []
    daily_status: List[RoomDailyStatus]


class MonthlyGridRequest(BaseModel):
    month: str = Field(..., pattern=r'^\d{4}-\d{2}$', description="Format: YYYY-MM")
    room_type_ids: Optional[List[UUID]] = None
    floors: Optional[List[int]] = None
    include_blocked: bool = True
    include_maintenance: bool = True


class MonthlyGridResponse(BaseModel):
    month: str
    days_in_month: int
    rooms: List[RoomGridData]
    summary: Dict[str, Any]
    occupancy_stats: List[Dict[str, Any]]


# Unassigned Bookings
class UnassignedBooking(BaseModel):
    booking_id: UUID
    booking_code: str
    guest_name: str
    phone: Optional[str] = None
    email: Optional[str] = None
    check_in_date: date
    check_out_date: date
    check_in_time: Optional[time] = None
    check_out_time: Optional[time] = None
    room_type: str
    room_type_id: Optional[UUID] = None
    adults: int = 1
    children: int = 0
    infants: int = 0
    total_guests: int = 1
    extra_persons: int = 0
    extra_person_charge: float = 0
    extra_bed: bool = False
    extra_bed_charge: float = 0
    special_requests: Optional[str] = None
    internal_notes: Optional[Dict[str, Any]] = None
    is_vip: bool = False
    hours_until_checkin: Optional[float] = None
    alert_level: AlertSeverity
    total_amount: Decimal
    paid_amount: Decimal
    booking_status: str = 'confirmed'


class UnassignedBookingsResponse(BaseModel):
    alerts: List[UnassignedBooking]
    summary: Dict[str, int]
    recommendations: List[str] = []


# Bulk Operations
class BulkResolveAlertsRequest(BaseModel):
    alert_ids: List[UUID]
    action: str = Field(..., description="auto_assign, manual_assign, dismiss")
    notes: Optional[str] = None
    assignment_strategy: Optional[AssignmentStrategy] = None


class BulkResolveAlertsResponse(BaseModel):
    resolved_count: int
    failed_count: int
    results: List[Dict[str, Any]]
    errors: List[str] = []


# Allocation Rules
class AllocationRuleBase(BaseModel):
    rule_name: str
    rule_type: str
    conditions: Dict[str, Any]
    actions: Dict[str, Any]
    priority: int = 0
    is_active: bool = True


class AllocationRuleCreate(AllocationRuleBase):
    created_by: Optional[UUID] = None


class AllocationRuleUpdate(BaseModel):
    rule_name: Optional[str] = None
    conditions: Optional[Dict[str, Any]] = None
    actions: Optional[Dict[str, Any]] = None
    priority: Optional[int] = None
    is_active: Optional[bool] = None


class AllocationRuleResponse(AllocationRuleBase):
    id: UUID
    created_by: Optional[UUID] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Guest Preferences
class GuestRoomPreferencesBase(BaseModel):
    customer_id: UUID
    preferred_room_types: List[UUID] = []
    preferred_floors: List[int] = []
    preferred_features: List[str] = []
    avoid_floors: List[int] = []
    avoid_features: List[str] = []
    avoid_rooms: List[str] = []
    accessibility_needs: Dict[str, Any] = {}
    special_requests: Optional[str] = None
    priority_level: int = Field(5, ge=1, le=10)


class GuestRoomPreferencesCreate(GuestRoomPreferencesBase):
    pass


class GuestRoomPreferencesUpdate(BaseModel):
    preferred_room_types: Optional[List[UUID]] = None
    preferred_floors: Optional[List[int]] = None
    preferred_features: Optional[List[str]] = None
    avoid_floors: Optional[List[int]] = None
    avoid_features: Optional[List[str]] = None
    avoid_rooms: Optional[List[str]] = None
    accessibility_needs: Optional[Dict[str, Any]] = None
    special_requests: Optional[str] = None
    priority_level: Optional[int] = Field(None, ge=1, le=10)


class GuestRoomPreferencesResponse(GuestRoomPreferencesBase):
    id: UUID
    last_stayed_room: Optional[str] = None
    total_stays: int
    satisfaction_score: Optional[int] = None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


# Optimization Reports
class OptimizationMetrics(BaseModel):
    average_assignment_time: str
    auto_assignment_rate: float
    manual_assignment_rate: float
    room_change_rate: float
    unassigned_alerts: int
    late_assignments: int


class RoomUtilization(BaseModel):
    room: str
    nights: int
    utilization: float
    revenue: Decimal
    avg_rate: Decimal


class OptimizationRecommendation(BaseModel):
    type: str
    message: str
    priority: str
    action_required: bool = False


class OptimizationReportResponse(BaseModel):
    period: Dict[str, date]
    metrics: OptimizationMetrics
    room_utilization: Dict[str, List[RoomUtilization]]
    recommendations: List[OptimizationRecommendation]
    trends: Dict[str, Any] = {}


# Available Rooms
class AvailableRoomInfo(BaseModel):
    room_id: UUID
    room_number: str
    room_type_id: UUID
    room_type: str
    floor: int
    features: List[str] = []
    base_rate: Decimal
    current_status: str
    availability_score: Optional[float] = None
    is_accessible: bool = False
    last_cleaned: Optional[datetime] = None
    maintenance_notes: Optional[str] = None


class AvailableRoomsRequest(BaseModel):
    check_in_date: date
    check_out_date: date
    room_type_id: Optional[UUID] = None
    guest_count: Optional[int] = None
    accessibility_required: bool = False
    features_required: List[str] = []
    exclude_rooms: List[UUID] = []
    booking_id: Optional[UUID] = None  # Include room assigned to this booking


class AvailableRoomsResponse(BaseModel):
    available_rooms: List[AvailableRoomInfo]
    total_available: int
    filters_applied: Dict[str, Any]
    suggestions: List[str] = []


# Room Assignment History
class AllocationHistoryResponse(BaseModel):
    id: UUID
    allocation_id: UUID
    booking_id: UUID
    action: str
    previous_room_id: Optional[UUID] = None
    new_room_id: Optional[UUID] = None
    previous_room_number: Optional[str] = None
    new_room_number: Optional[str] = None
    previous_dates: Optional[str] = None
    new_dates: Optional[str] = None
    previous_status: Optional[str] = None
    new_status: Optional[str] = None
    price_adjustment: Optional[Decimal] = None
    changed_by: Optional[UUID] = None
    change_reason: Optional[str] = None
    changed_at: datetime

    class Config:
        from_attributes = True


# Statistics and Analytics
class AllocationStatistics(BaseModel):
    total_allocations: int
    by_status: Dict[AssignmentStatus, int]
    by_type: Dict[AssignmentType, int]
    average_lead_time_hours: float
    room_utilization_rate: float
    conflict_resolution_time: float
    guest_satisfaction_impact: Optional[float] = None


class DailyAllocationSummary(BaseModel):
    date: date
    total_arrivals: int
    total_departures: int
    rooms_assigned: int
    rooms_available: int
    occupancy_rate: float
    unassigned_critical: int
    unassigned_warning: int
    conflicts_resolved: int